src/cryptonote_core/cryptonote_core.cpp

```diff
@@ -134,6 +135,14 @@ namespace cryptonote
   void core::stop()
   {
     m_blockchain_storage.cancel();
+
+    boost::lock_guard<boost::mutex> lock(m_update_mutex);
+    tools::download_async_handle handle = m_update_download;
```

Comment:
Ok, just saw your comment in the other PR in my inbox. I didn't even notice the data race at first actually. Its still a race because the callback could be invoked right as this code block is running.

The fix is to lock, copy handle, release, then call cancel. The callback would also lock.